Seafile: service and refactor

h2horn opened the pull request:
###### Motivation for this change

Seafile service, running a full seafile server.
- To build the seafile server, i had to change the existing seafile-shared package and renamed it to one seafile package.
- I updated all versions to 5.1.2-server tag, seafile-client (version 5.0.7) builds fine
- added a bunch of django python packages for seahub

Service file could use some more options:
- all databases are sqlite (seafile defaults)
- configured to run in fastcgi mode behind Nginx/Apache (see http://manual.seafile.com/deploy/deploy_with_nginx.html)
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

h2horn committed:
ccnet: 5.0.7 -> 5.1.2

- use server release tag, clients build fine
- fix license to gpl2
- use patchShebangs instead of sed

h2horn committed:
seafile: 5.0.7 -> 5.1.2, init seafile server

- enable server components
- replaces seafile-shared
- use own version of libevhtp (older version, cmake flags)

h2horn committed:
seafile-client: fix seafile dependency

- use new seafile package
- delete obsolete seafile-shared

h2horn committed:
seahub: init 5.1.2

- seafile web frontend and python packages

h2horn committed:
seafile service: init at 5.1.2

- seafile server with seahub

h2horn committed:
seafile: wrap binaries, fix sandbox build

groxxda commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -0,0 +1,229 @@
+      type = types.string;
+      default = "127.0.0.1";
+      description = ''
+        The tcp port used by ccnet.
```
"port"

groxxda commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -0,0 +1,229 @@
+        ExecStart = "${pkgs.seafile-seahub}/manage.py runfcgi host=127.0.0.1 port=8000 pidfile=${runDir}/seahub.pid outlog=${logDir}/seahub_access.log errlog=${logDir}/seahub_error.log";
+        PIDFile = "${runDir}/seahub.pid";
+      };
+      preStart = ''
```
Looks to me as if this needs `PermissionsStartOnly=true`

Cornu replied:
Runs fine without `PermissionsStartOnly=true` with different systemd services on my server. `RuntimeDirectory` is created with right permissions and owner specified in `User`.

joachifm replied:
@Cornu what if `dataDir` is left at its default and does not exist? Seems like something is misconfigured if an unprivileged process can create directories under `/var/lib` at will.

Cornu replied:
You're right. `dataDir` needs root permissions to be created. I didn't test without `dataDir` already existing.
I will fix it, so only `ccnet` as the first service to be started and creating `dataDir` should need
`PermissionsStartOnly=true`.

groxxda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -9029,6 +9029,88 @@ in modules // {
+    name = "django-statici18n-1.1.5";
+
+    src = pkgs.fetchurl {
+      url = "https://pypi.python.org/packages/source/d/django-statici18n/${name}.tar.gz";
```
you can use `mirror://pypi/d/...`

groxxda commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -0,0 +1,229 @@
+        User = "seafile";
+        Group = "seafile";
+        RuntimeDirectory = "seafile";
+        WorkingDirectory = pkgs.seafile-seahub;
```
I'm not sure whether setting WorkingDirectory to something inside the nix store is a good idea. Could be ok though.

joachifm replied:
Seems a little fragile to me, would it make sense to instead set it to the data dir? If the service expects the files in the store path to be in cwd, they could be copied at `preStart`.

Cornu replied:
I will change this, but why is this fragile?
Would it make a difference to link the directory into `dataDir` and use the link?
The `seahub` service is Django python app and needs to run in the `cwd`.

joachifm replied:
@Cornu I don't know all the potential states of this particular software, but it's quite natural for a daemon to be coded with the expectation of being able to write to the current working directory.

groxxda commented:
Just skimmed across the diff. See inline annotations.

joachifm commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -0,0 +1,229 @@
+  ###### Interface
+
+  options.services.seafile = {
+    enable = mkEnableOption ''
```
I believe this will render poorly. The `mkEnableOption` helper expects simply the name of the service; if you wish to provide a longer form description, you must override it (as in `mkEnableOption "foo" // { description = "My long description"; }`).

joachifm commented on lib/maintainers.nix:
```diff
@@ -160,6 +160,7 @@
   hiberno = "Christian Lask <hiberno@hiberno.net>";
   hinton = "Tom Hinton <t@larkery.com>";
   hrdinka = "Christoph Hrdinka <c.nix@hrdinka.at>";
+  hhhorn = "Hans-Harro Horn <h.h.horn@gmail.com>";
```
Would you mind using your github handle here?

danbst commented on nixos/modules/services/networking/seafile.nix:
```diff
@@ -0,0 +1,229 @@
+    enable = mkEnableOption ''
+      Enable Seafile Server.
+
+      For further information see [Seafile Manual](http://manual.seafile.com/).
```
The link should be in docbook style

FRidh commented:
@Cornu can you make separate commits for each package. 
With the Python packages you will most likely have to explicitly disable tests with `doCheck = false;` because Django packages often don't come with tests, and when they do, they typically don't run out of the box and are hard to set up.

Cornu commented:
@FRidh do you mean for each Python package?

I will push new commits this weekend, i'm a little busy right now.

FRidh commented:
Indeed, for each Python package.

kampfschlaefer committed:
seafile: Start a test

Start with a test to proof the seafile-server module works.

Also fix the first errors shown by the test.

kampfschlaefer committed:
seafile: 5.1.2 -> 5.1.4

The tests show that 5.1.2 has an error in seaf-cli which is fixed in the
latest 5.1.x version.

kampfschlaefer commented:
I started writing a test for this module in https://github.com/Cornu/nixpkgs/pull/1

kampfschlaefer committed:
seafile: run seahub-fcgi on private address

It is meant to be run behind an nginx proxy.

kampfschlaefer committed:
seafile: typo fix

kampfschlaefer committed:
seafile-test: Test with nginx-proxy and two clients

Extend the tests:
- nginx as proxy before the seahub-fcgi
- one client to create a library and place a file in it
- second client to download that library and check the file in it

uwap commented:
@cornu Any updates on this?

h2horn committed:
Merge pull request #1 from kampfschlaefer/cornu-seafile

Add test for seafile-module

Cornu commented:
@uwap Hey, i'm not running seafile at the moment, so won't work on this anytime soon.
I would however merge pull requests and help with improvements, so feel free to help.

@kampfschlaefer Thanks for the module test

So open issues are:
* update to current version of seafile
* split python packages into seperate commits

jeaye commented:
Having a Seafile service in NixOS would be superb; thanks for this work!

Infinisil commented:
Closing this for now, doesn't look like this will get finished anytime soon.
